### PR TITLE
Simple env

### PR DESCRIPTION
**Pull request by DT6A**
Implementing an environment with one hider and one seeker

### CODE REVIEW

**Comment by DT6A** [comment]
Rewards during training
image

File: MAHS/Assets/SimpleEnv/Editor/ViewFieldEditor.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Editor/ViewFieldEditor.cs:1]: Unused imports

File: MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
+﻿using System.Collections;
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs:1]: Unused imports

File: MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    BehaviorParameters behaviorParameters;
+    [HideInInspector]
+    public Team team;
+    void Start()
+    {
+        rBody = GetComponent<Rigidbody>();
+        vf = GetComponent<ViewField>();
+        behaviorParameters = GetComponent<BehaviorParameters>();
+        
+        if (behaviorParameters.TeamId == (int) Team.Hider)
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs:34]: can be simplified:
`team = behaviorParameters.TeamId == (int) Team.Hider ? Team.Hider : Team.Seeker;`

File: MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            team = Team.Hider;
+        }
+        else
+        {
+            team = Team.Seeker; 
+        }
+
+        this.MaxStep = area.episodeLength;
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs:43]: `this` qualifier is usually redundant

File: MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    
+    public override void OnEpisodeBegin()
+    {
+        this.rBody.angularVelocity = Vector3.zero;
+        this.rBody.velocity = Vector3.zero;
+        this.transform.Rotate(0,Random.value * 360 - 180f, 0);
+        
+        if (team == Team.Hider)
+        {
+            this.transform.localPosition = new Vector3(2 * (Random.value * 4.25f + 0.25f), 0.5f, 2 * (-Random.value * 4.25f - 0.25f));
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs:56]: Let's extract all these magic constants so it's easy to change them when the time comes.

File: MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        sensor.AddObservation(transform.localPosition.x);
+        sensor.AddObservation(transform.localPosition.z);
+        // Agent velocity
+        sensor.AddObservation(rBody.velocity.x);
+        sensor.AddObservation(rBody.velocity.z);
+
+        Quaternion opponentInfo = new Quaternion(0, 0, 0, 0);
+        foreach (GameObject obj in vf.collectVisibleObjects())
+        {
+            if (obj.CompareTag("Hider") || obj.CompareTag("Seeker"))
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs:82]: I'd invert this `if` to improve readability

Comment (ktisha) [reply MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs:82]: `if (!obj.CompareTag("Hider") && !obj.CompareTag("Seeker")) continue;`

File: MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                opponentInfo.y =  obj.transform.localPosition.z;
+                // Opponent velocity
+                opponentInfo.z = otherAgentRigidbody.velocity.x;
+                opponentInfo.w = otherAgentRigidbody.velocity.z;
+                break;
+            }
+        }
+        
+        sensor.AddObservation(opponentInfo.x);
+        sensor.AddObservation(opponentInfo.y);
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/AgentScript.cs:101]: Why do we need opponents `y` and `w` in observations?

File: MAHS/Assets/SimpleEnv/Scripts/TrainingAreaScript.cs [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    public Transform box2;
+    public void ResetEnv()
+    {
+        box1.transform.localPosition = new Vector3(Random.value * 4.25f + 0.25f, 0.5f, -Random.value * 4.25f - 0.25f);
+        box2.transform.localPosition = new Vector3(Random.value * 4.25f + 0.25f, 0.5f, -Random.value * 4.25f - 0.25f);
+    }
+
+    void PreStep(int i)
+    {
+        float hidersReward = 1.0f;
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/TrainingAreaScript.cs:34]: Let's simplify it as following:
`var hidersReward = isAnyHiderSeen ? -1.0f : 1.0f;`

File: MAHS/Assets/SimpleEnv/Scripts/TrainingAreaScript.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        if (isAnyHiderSeen)
+            hidersReward = -1.0f;
+        
+        foreach (AgentScript agent in agents)
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/TrainingAreaScript.cs:38]: There are LINQ expressions in c#:
```
foreach (var agent in agents.TakeWhile(agent => agent.StepCount >= preparingPhaseLength))
        {
            if (agent.team == AgentScript.Team.Hider)
                agent.AddReward(hidersReward);
            else
                agent.AddReward(-hidersReward);
        }
```

File: MAHS/Assets/SimpleEnv/Scripts/ViewField.cs [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        List<GameObject> objects = new List<GameObject>();
+
+        Collider[] colliders = Physics.OverlapSphere(transform.position, viewRadius);
+        for (int i = 0; i < colliders.Length; i++)
+        {
+            Transform trans = colliders[i].transform;
+            Vector3 dirToTarget = (trans.position - transform.position).normalized;
+
+            if (!colliders[i].CompareTag("Wall")  && Vector3.Angle(transform.forward, dirToTarget) < viewAngle / 2.0f)
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/ViewField.cs:21]: Let's extract the tag name `Wall` somewhere so we can easily change it.

File: MAHS/Assets/SimpleEnv/Scripts/ViewField.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+            {
+                float dist = Vector3.Distance(trans.position, transform.position);
+                RaycastHit hit;
```
Comment (ktisha) [inline MAHS/Assets/SimpleEnv/Scripts/ViewField.cs:24]: This variable introduction can be inlined:
`if (Physics.Raycast(transform.position, dirToTarget, out var hit, dist))`

**Pull request merged**